File: java/springmvc/src/main/java/com/lacunasoftware/pkisuite/controller/ServerFilesController.java
package com.lacunasoftware.pkisuite.controller;

import com.lacunasoftware.pkisuite.model.ServerFile;
import com.lacunasoftware.pkisuite.model.ServerFilesModel;
import com.lacunasoftware.pkisuite.util.SampleDocs;
import com.lacunasoftware.pkisuite.util.StorageMock;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

@Controller
@RequestMapping("/server-files")
public class ServerFilesController {

	/**
	 * GET /server-files
	 *
	 * This method will render a page to show the available server files to be used on the sample
	 * referred by "rc" parameter.
	 */
	@GetMapping
	public String get(
		@RequestParam String rc,
		@RequestParam String op,
		Model model
	) throws IOException {

		List<ServerFile> availableFiles = new ArrayList<ServerFile>();
		boolean isCmsCoSign = false;

		switch (op) {
			case "cosignCms":
				isCmsCoSign = true;
				availableFiles.add(new ServerFile(SampleDocs.CMS_SIGNED_ONCE, "A sample CMS file that was signed once."));
				availableFiles.add(new ServerFile(SampleDocs.CMS_SIGNED_TWICE, "A sample CMS file that was signed twice."));
				break;
			case "cosignPdf":
			case "printerFriendly":
				availableFiles.add(new ServerFile(SampleDocs.PDF_SIGNED_ONCE, "A sample PDF that was signed just once."));
				availableFiles.add(new ServerFile(SampleDocs.PDF_SIGNED_TWICE, "A sample PDF that was signed twice."));
				break;
			case "signCms":
			case "signPdf":
				availableFiles.add(new ServerFile(SampleDocs.SAMPLE_PDF, "A sample PDF document (size: 25kb) with no signatures."));
				break;
			default:
				throw new FileNotFoundException();
		}

		// It is up to your application's business logic to determine which server documents
		// to be available for the signature.
		ServerFilesModel serverFilesModel = new ServerFilesModel();
		serverFilesModel.setIsCmsCoSign(isCmsCoSign);
		serverFilesModel.setAvailableFiles(availableFiles);

		model.addAttribute("model", serverFilesModel);
		return "server-files/index";
	}

	/**
	 * POST /server-files
	 *
	 * This method is called when the user chooses a sample file. It will get the file's content and
	 * will store it on a temporary folder using our StorageMock class.
	 */
	@PostMapping
	public String post(
		@RequestParam String rc,
		@ModelAttribute ServerFilesModel model
	) throws IOException {

		Path path = StorageMock.getSampleDocPath(model.getChosenFileId());
		String filename = StorageMock.getSampleDocName(model.getChosenFileId());

		// Copy file to the temporary folder, where the upload files are stored.
		String fileId;
		try (InputStream fileStream = new FileInputStream(path.toFile())) {
			fileId = StorageMock.store(fileStream, "pdf", filename);
		}

		// Only the REST PKI sample needs to pass this file as "cosign" variable when cosigning a
		// CMS file.
		if (rc.equals("cades-signature-rest") && model.getIsCmsCoSign()) {
			return String.format("redirect:/%s?cosign=%s", rc, fileId);
		}

		return String.format("redirect:/%s?fileId=%s", rc, fileId);
	}
}

File: ruby/rails/app/controllers/server_files_controller.rb
class ServerFilesController < ApplicationController

  # GET /server-files
  # This action will render a page to show the available server files to be used
  # on the sample referred by "rc" parameter.
  def index
    @rc = params[:rc]
    @op = params[:op]

    @available_files = []
    @is_cms_cosign = false

    case @op
    when 'cosign-cms'
      @is_cms_cosign = true
      @available_files.append(id: :cms_signed_once, description: 'A sample CMS file that was signed once.')
      @available_files.append(id: :cms_signed_twice, description: 'A sample CMS file that was signed twice.')
    when 'cosign-pdf'
      @available_files.append(id: :pdf_signed_once, description: 'A sample PDF that was signed just once.')
      @available_files.append(id: :pdf_signed_twice, description: 'A sample PDF that was signed twice.')
    when 'printer-friendly'
      @available_files.append(id: :pdf_signed_once, description: 'A sample PDF that was signed just once.')
      @available_files.append(id: :pdf_signed_twice, description: 'A sample PDF that was signed twice.')
    when 'sign-cms'
      @available_files.append(id: :sample_pdf, description: 'A sample PDF document (size: 25kb) with no signatures.')
    when 'sign-pdf'
      @available_files.append(id: :sample_pdf, description: 'A sample PDF document (size: 25kb) with no signatures.')
    else
      raise Exception('Invalid operation')
    end
    # Is it up to your application's business logic to determine which server
    # documents to be available for the signature.
  end

  # POST /server-files
  # This method is called when the user chooses a sample file. It will get the
  # file's content and will store it on a temporary folder using our StorageMock
  # class.
  def action

    path = get_sample_doc_path(params[:chosen_file_id])
    filename = get_sample_doc_name(params[:chosen_file_id])

    # Copy file to the temporary folder, where the upload files are stored.
    file_id = nil
    if params[:is_cms_cosign] == 'false'
      File.open(path, 'rb') do |file|
        file_id = store(file.read, 'pdf', filename)
      end
    elsif params[:is_cms_cosign] == 'true'
      File.open(path, 'rb') do |file|
        file_id = store(file.read, 'p7s', filename)
      end
    end

    # Only the REST PKI sample needs to pass this file as "cosign" variable when
    # cosigning a CMS file.
    if (params[:rc] == 'cades-signature-rest') && (params[:is_cms_cosign] == 'true')
      redirect_to "/#{params[:rc]}/cosign/#{file_id}"
    else
      redirect_to "/#{params[:rc]}/#{file_id}"
    end
  end
end
